std-bits/table/src/main/java/org/enso/table/write/ExcelWriter.java

```diff
@@ -127,47 +126,99 @@ public static void writeTableToRange(Workbook workbook, ExcelRange range, boolea
+    headers = headers != ExcelHeaders.HeaderBehavior.INFER ? headers :
+        shouldWriteHeaders(sheet, expanded.getTopRow(), expanded.getLeftColumn(), expanded.getRightColumn());
 
-      if (headers == ExcelHeaders.HeaderBehavior.INFER) {
-        headers = shouldWriteHeaders(sheet, expanded.getTopRow(), expanded.getLeftColumn(), expanded.getRightColumn())
-            ? ExcelHeaders.HeaderBehavior.USE_FIRST_ROW_AS_HEADERS
-            : ExcelHeaders.HeaderBehavior.EXCEL_COLUMN_NAMES;
-      }
+    if ((existingDataMode == ExistingDataMode.APPEND_BY_NAME || existingDataMode == ExistingDataMode.APPEND_BY_INDEX) &&
+        rangeIsNotEmpty(workbook, expanded, false, sheet)) {
+      appendRangeWithTable(workbook, range, existingDataMode, table, rowLimit, headers, sheet, expanded);
+    } else {
+      updateRangeWithTable(workbook, expanded, range.isSingleCell(), existingDataMode, table, rowLimit, headers, sheet);
+    }
+  }
+
+  /**
+   * Creates an empty workbook.
+   * @param xls_format specifies whether the file is in Excel Binary Format (95-2003 format).
+   * @return a {@link Workbook} containing the specified data.
+   */
+  public static Workbook createWorkbook(boolean xls_format) {
+    return xls_format ? new HSSFWorkbook() : new XSSFWorkbook();
+  }
 
-      // Expand to cover required size
-      int rowCount = (headers == ExcelHeaders.HeaderBehavior.USE_FIRST_ROW_AS_HEADERS ? 1 : 0) + table.rowCount();
-      if (expanded.getColumnCount() < table.getColumns().length || expanded.getRowCount() < rowCount) {
-        expanded = new ExcelRange(
-            expanded.getSheetName(),
-            expanded.getLeftColumn(),
-            expanded.getTopRow(),
-            Math.max(expanded.getRightColumn(), expanded.getLeftColumn() + table.getColumns().length - 1),
-            Math.max(expanded.getBottomRow(), expanded.getTopRow() + rowCount - 1));
+  private static void appendRangeWithTable(Workbook workbook, ExcelRange range, ExistingDataMode existingDataMode, Table table, Long rowLimit, ExcelHeaders.HeaderBehavior headers, ExcelSheet sheet, ExcelRange expanded)
+      throws RangeExceededException, ExistingDataException, ColumnNameMismatchException, ColumnCountMismatchException {
+    // Map Table
+    Table mappedTable = switch (existingDataMode) {
+      case APPEND_BY_INDEX -> ColumnMapper.MapColumnsByIndex(table, expanded.getColumnCount());
+      case APPEND_BY_NAME -> {
+        if (headers == ExcelHeaders.HeaderBehavior.EXCEL_COLUMN_NAMES) {
+          throw new IllegalArgumentException("Cannot append by name when headers are not present in the existing data.");
+        }
+        String[] currentHeaders = sheet.get(expanded.getTopRow()).getCellsAsText(expanded.getLeftColumn(), expanded.getRightColumn());
+        yield ColumnMapper.MapColumnsByName(table, currentHeaders);
       }
+      default ->
+          throw new IllegalArgumentException("The existing data mode '" + existingDataMode + "' is not supported.");
```

Comment:
This message suggests that this mode is not supported by our Excel Write at all. But if I understand correctly it is actually an internal error where this internal helper function has been called with an unexpected mode, right? If this is indeed the case, I think it may be good to mark this somehow with an "Internal Error" prefix, to let the user know that this is a bug and not mislead them that some functionality is not supported.